tools/fp20compiler/main.cpp

```diff
@@ -37,25 +37,69 @@ char * myin = 0;
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
-
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
-
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
-    }
-
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
-    }
+    // Move to start of file
+    char* cursor = ns;
+
+    // Remember where the current shader starts
+    char* shader_start = NULL;
+
+    // Loop over each shader (between two shader start tokens)
+    do {
+
+        // Try to find a shader start token candidate to mark the end
+        char* shader_end = strstr(cursor, "!!");
+
+        // Make sure we found a token candidate
+        if (shader_end != NULL) {
+
+            // Only accept token at beginning of file, or beginning of line
+            if ((shader_end != ns) && (shader_end[-1] != '\n')) {
+                cursor = shader_end + 1;
+                continue;
+            }
+
+        }
```

Comment:
nit: This should probably have been a function: `shader_end = find_line_start(ns, cursor, "!!");`

Reply:
I've done this now. The function name is slightly confusing, but still okay.